### ISSUE TITLE
getExecutable loops slow down Pimcore at Windows Systems

### ISSUE THREAD

**Issue by Cruiser13** [pimcore/pimcore#1075]
# Bug Report
I know that Windows is not officially supported but we have quite a few intranet solutions based on pimcore which are running on a Windows System so I give it a try...

With pimcore 4.X there have been changes to the path configuration in the pimcore settings. We had some issues which have been resolved now and we updated to the latest pimcore 4.X on the windows systems. After that, we noticed a huge performance loss in the admin CP. The assets will load ~50 seconds after the whole admin interface is available. Sometimes they won't load at all and this error is thrown:
![admin-asset-error](https://cloud.githubusercontent.com/assets/2771909/20598176/1e5b9c48-b248-11e6-8103-3035dcb65795.png)

The problem gets worse if we add more software to the PATH system environment variable in Windows. It looks like pimcore needs too much time for these loops on windows: https://github.com/pimcore/pimcore/blob/master/pimcore/lib/Pimcore/Tool/Console.php#L90

We worked around this problem by adding code like this to the console.php in line 67 for software which is directly available:
```php
		if($name == 'composer'){
			return 'composer';
		}
```
But that is no upgrade-safe solution and it is error prone as well. And it still does not speed up things so much because there are checks for software which is not installed (because it's not available or not needed).

It'd be great to get a different solution - to specify the path to certain software directly for example like it has been done in the 3.X series.

**Comment by bbak** [comment]
Is there a reason why this method doesn't just utilize `which` und Un*x and `where.exe` on Windows to find executables but instead loops through path's and tries to create processes?

**Comment by Cruiser13** [comment]
I don't know if that is a proper solution. Alternatively we could add optional config parameters to system.php which could be checked for the application before going into the loops. We could define proper paths for Windows in this case.

**Comment by bbak** [comment]
Of course it's not 100% error-safe, but it's quite unlikely that `which` or `where.exe` find the wrong executable (e.g. `soffice.exe`) in the OS path. And if yes, I'd say, the System has a general Setup/Installation-Problem.
The Performance gain would be quite huge. 

@drietsch What do you think?

**Comment by brusch** [comment]
Customizing the path to external programs is now possible by using DI, see #1007

**Comment by brusch** [comment]
In general the idea is that the PATH variable in system settings can overwrite OS-level settings, eg. when you have a dedicated installation of LibreOffice for your Pimcore installation. In my opinion this makes sense, so we shouldn't change the order of discovering the real path to external programs. 

@Cruiser13 what exactly takes so long on Windows system? is it `$process->run();` ? 
Does ist help to change the code to: 
```php
$process = new Process($executablePath . " " . $option);
$process->setTimeout(.2);
$process->run();
```

**Comment by Cruiser13** [comment]
@brusch yes, all the processes being run slows down the system. Your solution does speed up the loading time for the requirements check in our environment. It went from ~11 seconds to ~5 seconds.

Customizing the paths to the external programs directly will also be a great solution if pimcore does add them to the $executableCache in this case so they'll not cause all these foreach loops.

**Comment by brusch** [comment]
That should be fixed already by abbe44d414dbe262ad24afc3fffae16406472ea0. Pimcore now utilizies `which` and `where` to determine the right path. 
Additionally you have still the possibility to manually specify the path to the executables: https://github.com/pimcore/demo-basic/blob/master/app/config/parameters.example.yml#L9
